src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferConstCharOverConstUnitString.Fixer.cs

```diff
@@ -0,0 +1,111 @@
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public class PreferConstCharOverConstUnitStringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(PreferConstCharOverConstUnitStringAnalyzer.RuleId);
+
+        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            Document doc = context.Document;
+            CancellationToken cancellationToken = context.CancellationToken;
+            SyntaxNode root = await doc.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            if (root.FindNode(context.Span) is SyntaxNode expression)
+            {
+                context.RegisterCodeFix(
+                    CodeAction.Create(
+                        title: MicrosoftNetCoreAnalyzersResources.PreferConstCharOverConstUnitStringInStringBuilderMessage,
+                        createChangedDocument: async c =>
+                        {
+                            SemanticModel semanticModel = await doc.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                            var operation = semanticModel.GetOperationWalkingUpParentChain(expression, cancellationToken);
+                            if (operation is IVariableDeclaratorOperation variableDeclaratorOperation)
+                            {
+                                return await HandleVariableDeclarator(variableDeclaratorOperation, doc, root, cancellationToken).ConfigureAwait(false);
+                            }
+                            else if (operation is IArgumentOperation argumentOperation)
+                            {
+                                if (argumentOperation.Value is ILiteralOperation argumentLiteral)
+                                {
+                                    return await HandleStringLiteral(argumentLiteral, doc, root, cancellationToken).ConfigureAwait(false);
+                                }
+                            }
+                            return doc;
+                        },
+                        equivalenceKey: "PreferConstCharOverConstUnitStringInStringBuilderAppend"),
+                    context.Diagnostics);
+            }
+        }
+
+        private static async Task<Document?> HandleStringLiteral(ILiteralOperation argumentLiteral, Document doc, SyntaxNode root, CancellationToken cancellationToken)
+        {
+            if (argumentLiteral.ConstantValue.HasValue && argumentLiteral.ConstantValue.Value is string unitString && unitString.Length == 1)
+            {
+                DocumentEditor editor = await DocumentEditor.CreateAsync(doc, cancellationToken).ConfigureAwait(false);
```

Comment:
We should register a code fix only if you are certain to reach a code path where document will be changed.

Reply:
Basically, I am suggesting use the approach similar to the one taken here: https://github.com/dotnet/roslyn-analyzers/blob/6b16c9c25464f5f77d33da7a62f803857d4d3d6a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseAsSpanInsteadOfRangeIndexer.Fixer.cs#L51. Have all the logic to find the relevant operation with validations to be done before invoking `RegisterCodeFix`

Reply:
The `if` blocks here will always succeed because the analyzer verifies everything before calling the fixer like you suggested. I removed the `if` blocks now.